langchain-core/src/callbacks/manager.ts

```diff
@@ -1256,9 +1256,14 @@ export class CallbackManager
       if (tracingV2Enabled) {
         // handoff between langchain and langsmith/traceable
         // override the parent run ID
-        callbackManager._parentRunId =
-          LangChainTracer.getTraceableRunTree()?.id ??
-          callbackManager._parentRunId;
+        const implicitRunTree = LangChainTracer.getTraceableRunTree();
+        if (implicitRunTree && callbackManager._parentRunId === undefined) {
+          callbackManager._parentRunId = implicitRunTree.id;
+          const tracerV2 = callbackManager.handlers.find(
+            (handler) => handler.name === "langchain_tracer"
+          ) as LangChainTracer;
+          tracerV2?.updateFromRunTree(implicitRunTree);
```

Comment:
Previously this was always setting `callbackManager._parentRunId` to the implicit traceable value, so runs would get flattened

Previously this only applied when a `LangChainTracer` was instantiated automatically if it were missing, but my fix in #8849 for cases where a tracer is passed manually caused this reparenting logic to apply even if the tracer were already present